### ISSUE TITLE
IXLWorksheet.CopyTo works as expected in a WPF app, not in a WinForm app

### ISSUE THREAD

**Issue by glgdu** [ClosedXML/ClosedXML#1188]
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

If you are logging a possible bug or feature request, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**

v4.0.30319

**What is the current behavior?**

The following code:

var workbookSource = new XLWorkbook(sourceFile);
var wsSourceControl = workbookSource.Worksheet("Control");
wsSourceControl.CopyTo(workbookTarget, "Control", 1);

works as expected when I first used it in a WPF app (Visual Studio 2017 Community v14.1.2). When I use the same code in a WinForm app, it seems to enter a loop, and ultimately triggers the "ContextSwitchDeadlock" exception:
Message=Managed Debugging Assistant 'ContextSwitchDeadlock' : 'The CLR has been unable to transition from COM context 0xd27270 to COM context 0xd271b8 for 60 seconds. The thread that owns the destination context/apartment is most likely either doing a non pumping wait or processing a very long running operation without pumping Windows messages. This situation generally has a negative performance impact and may even lead to the application becoming non responsive or memory usage accumulating continually over time. To avoid this problem, all single threaded apartment (STA) threads should use pumping wait primitives (such as CoWaitForMultipleHandles) and routinely pump messages during long running operations.'

**What is the expected behavior or new feature?**

I would expect the code to perform consistently irrespective of the application type being developed.

**Did this work in previous versions of our tool?  Which versions?**

No - only tried this version and the previous one though.

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
private void button1_Click(object sender, EventArgs e)
        {
            string sourceFile = "CopyTo test.xlsx";
            var workbookTarget = new XLWorkbook();
            var workbookSource = new XLWorkbook(sourceFile);
            var wsSourceControl = workbookSource.Worksheet("Control");
            wsSourceControl.CopyTo(workbookTarget, "Control", 1);
        }
```
- [X] I attached a sample spreadsheet.  (You can drag files on to this issue)

[CopyTo test.xlsx](https://github.com/ClosedXML/ClosedXML/files/3092024/CopyTo.test.xlsx)

**Comment by Pankraty** [comment]
The exception you are referring to is thrown by Visual Studio when it does not get the response in 60 seconds. This not necessarily indicates the infinite loop, sometimes it is just a long-running operation.

In your example, ClosedXML performs a heavy job of copying 16133 columns from one workbook to another. I don't know why it takes so long, and it is definitely worth investigation, but it has no relation to WinForms vs. WPF (it takes minutes to finish the copying in a console app as well).

**Comment by glgdu** [comment]
Thank you for your answer. You are right, it is not an infinite loop. But it takes 2 minutes 20 seconds to execute the sample code in the WPF app, and 35 minutes in the WinForm app. This is a big enough difference to make CopyTo unusable in WinForm - and as you say, it would be interesting to know the reason why this happens.

**Comment by Pankraty** [comment]
I was going to take a look why the copying works so slowly after the vacation, so I'll reopen it, if you don't mind.

**Comment by igitur** [comment]
Ok, sure.

**Comment by Pankraty** [comment]
Turned out, it is the combination of two factors that made the worksheet in this workbook to be copied so slowly.

First, there are more than 16 thousand columns initialized. Even though almost all of them are empty they still have to be copied.

Second, there are 1024 data validation rules scattered here and there (for example, did you know, there is one on a cell `WBR983048`? This looks absolutely random).

With two factors combined ClosedXML had to perform 1024 * 16K operations, which does not seem efficient.

I managed to improve the performance from ~23 minutes to ~1m40s but I see a few more issues regarding data validation rules. Eliminating them is going to speed up copying even better.

**Comment by glgdu** [comment]
Thank you very much for following up on this. The extent to which Excel input can be random will never stop surprising me. I was aware of the empty but initialised cells, not of the random validation rules.

May I ask if you find any indication of why performance is so different between WPF and Winform apps? That really does not make any sense to me. I kind of think that discovering the reason may be interesting beyond the scope of the immediate issue.

Are the improvements that you mention going to be included in a future release?

On 10 May 2019, at 5:37 pm, Aleksei <notifications@github.com<mailto:notifications@github.com>> wrote:


Turned out, it is the combination of two factors that made the worksheet in this workbook to be copied so slowly.

First, there are more than 16 thousand columns initialized. Even though almost all of them are empty they still have to be copied.

Second, there are 1024 data validation rules scattered here and there (for example, did you know, there is one on a cell WBR983048? This looks absolutely random).

With two factors combined ClosedXML had to perform 1024 * 16K operations, which does not seem efficient.

I managed to improve the performance from ~23 minutes to ~1m40s but I see a few more issues regarding data validation rules. Eliminating them is going to speed up copying even better.

**Comment by Pankraty** [comment]
Hi @glgdu ,

The difference between WPF and WinForms application seems strange for me too. But I am afraid I will not be able to investigate this in the near future - my vacation is over and I am going to have a very busy week (weeks?..).

Can you check the binaries from (this build)[https://ci.appveyor.com/project/ClosedXML/closedxml/builds/24486265/job/jriq2qiq168mq30s/artifacts], if the difference still noticeable?

Do both your applications use the same version of .NET framework? Which version?

As for future releases - it is usually up to @igitur to decide what and where to include in releases. I would suggest to include #1205 in 0.95 and #1206 in 0.96, as there were significant changes and 0.95 is already in beta state. But maybe both fixes will be postponed till 0.96.

**Comment by Pankraty** [comment]
I made some testing with different types of application (Windows Forms and WPF) and different target platforms (.NET Framework 4.0, .NET Framework 4.6, .NET Core 3.0) and did not notice any statistically significant difference between WinForms and WPF.

.NET Core version looks slightly faster, at average, but I am cautious in judgments because diversity in different runs of the same application was larger than the noticed difference between frameworks.
